llarp/path/path_context.hpp

```diff
@@ -25,149 +25,130 @@ namespace llarp
+  {
+    return lhs.operator==(rhs);
+  }
+
+  template <>
+  struct hash<llarp::path::TransitHopID>
+  {
+    size_t
+    operator()(const llarp::path::TransitHopID& obj) const noexcept
     {
-      explicit PathContext(Router* router);
+      return std::hash<llarp::PathID_t>{}(obj.path_id);
+    }
+  };
+}  // namespace std
 
-      /// called from router tick function
-      void
-      ExpirePaths(llarp_time_t now);
+namespace llarp::path
+{
+  using TransitHop_ptr = std::shared_ptr<TransitHop>;
 
-      void
-      PumpUpstream();
+  struct PathContext
+  {
+    explicit PathContext(Router* router);
 
-      void
-      PumpDownstream();
+    /// called from router tick function
+    void
+    ExpirePaths(llarp_time_t now);
 
-      void
-      AllowTransit();
+    void
+    AllowTransit();
 
-      void
-      RejectTransit();
+    void
+    RejectTransit();
 
-      bool
-      CheckPathLimitHitByIP(const IpAddress& ip);
+    bool
+    CheckPathLimitHitByIP(const IpAddress& ip);
 
-      bool
-      CheckPathLimitHitByIP(const std::string& ip);
+    bool
+    CheckPathLimitHitByIP(const std::string& ip);
 
-      bool
-      AllowingTransit() const;
+    bool
+    AllowingTransit() const;
 
-      bool
-      HasTransitHop(const TransitHopInfo& info);
+    bool
+    HasTransitHop(const TransitHopInfo& info);
 
-      void
-      PutTransitHop(std::shared_ptr<TransitHop> hop);
+    void
+    PutTransitHop(std::shared_ptr<TransitHop> hop);
 
-      HopHandler_ptr
-      GetByUpstream(const RouterID& id, const PathID_t& path);
+    Path_ptr
+    GetPath(const PathID_t& path_id);
 
-      bool
-      TransitHopPreviousIsRouter(const PathID_t& path, const RouterID& r);
+    bool
+    TransitHopPreviousIsRouter(const PathID_t& path, const RouterID& r);
 
-      TransitHop_ptr
-      GetPathForTransfer(const PathID_t& topath);
+    TransitHop_ptr
+    GetPathForTransfer(const PathID_t& topath);
 
-      HopHandler_ptr
-      GetByDownstream(const RouterID& id, const PathID_t& path);
+    std::shared_ptr<TransitHop>
+    GetTransitHop(const RouterID&, const PathID_t&);
 
-      std::optional<std::weak_ptr<TransitHop>>
-      TransitHopByInfo(const TransitHopInfo&);
+    PathSet_ptr
+    GetLocalPathSet(const PathID_t& id);
 
-      std::optional<std::weak_ptr<TransitHop>>
-      TransitHopByUpstream(const RouterID&, const PathID_t&);
+    using EndpointPathPtrSet = std::set<Path_ptr, ComparePtr<Path_ptr>>;
+    /// get a set of all paths that we own who's endpoint is r
+    EndpointPathPtrSet
+    FindOwnedPathsWithEndpoint(const RouterID& r);
 
-      PathSet_ptr
-      GetLocalPathSet(const PathID_t& id);
+    bool
+    HopIsUs(const RouterID& k) const;
 
-      using EndpointPathPtrSet = std::set<Path_ptr, ComparePtr<Path_ptr>>;
-      /// get a set of all paths that we own who's endpoint is r
-      EndpointPathPtrSet
-      FindOwnedPathsWithEndpoint(const RouterID& r);
+    void
+    AddOwnPath(PathSet_ptr set, Path_ptr p);
```

Comment:
And definitely `Path_ptr` too

Reply:
agreed, but leaving this one specifically for a future PR as many of the places it's used should be a reference instead of copying a shared_ptr, and many others will likely go away entirely.  I can make that change in this PR if you want though.